zkevm-circuits/src/pi_circuit.rs

```diff
@@ -0,0 +1,978 @@
+                        ),
+                        (
+                            TxFieldTag::CallerAddress,
+                            tx.from_addr.to_scalar().expect("tx.from too big"),
+                        ),
+                        (
+                            TxFieldTag::CalleeAddress,
+                            tx.to_addr.to_scalar().expect("tx.to too big"),
+                        ),
+                        (TxFieldTag::IsCreate, F::from(tx.is_create)),
+                        (
+                            TxFieldTag::Value,
+                            rlc(tx.value.to_le_bytes(), self.randomness),
+                        ),
+                        (TxFieldTag::CallDataLength, F::from(tx.call_data_len)),
+                        (
+                            TxFieldTag::TxSignHash,
+                            rlc(tx.tx_sign_hash, self.randomness),
+                        ),
+                    ] {
+                        config.assign_tx_row(
+                            &mut region,
+                            offset,
+                            i + 1,
+                            *tag,
+                            0,
+                            *value,
+                            &mut raw_pi_vals,
+                        )?;
+                        offset += 1;
+                    }
+                }
+                // Tx Table CallData
+                let mut calldata_count = 0;
+                for (i, tx) in self.public_data.txs.iter().enumerate() {
+                    for (index, byte) in tx.call_data.0.iter().enumerate() {
+                        assert!(calldata_count < MAX_CALLDATA);
+                        config.assign_tx_row(
+                            &mut region,
+                            offset,
+                            i + 1,
+                            TxFieldTag::CallData,
+                            index,
+                            F::from(*byte as u64),
+                            &mut raw_pi_vals,
+                        )?;
+                        offset += 1;
+                        calldata_count += 1;
+                    }
+                }
+                for _ in calldata_count..MAX_CALLDATA {
+                    config.assign_tx_row(
+                        &mut region,
+                        offset,
+                        0, // tx_id
+                        TxFieldTag::CallData,
+                        0,
+                        F::zero(),
+                        &mut raw_pi_vals,
+                    )?;
+                    offset += 1;
+                }
+
+                // rpi_rlc and rand_rpi cols
+                let (rpi_rand, rpi_rlc) =
+                    config.assign_rlc_pi(&mut region, self.rand_rpi, raw_pi_vals)?;
+
+                Ok(vec![
+                    rpi_rand,
+                    rpi_rlc,
+                    chain_id,
+                    state_root,
+                    prev_block_hash,
```

Comment:
The spec defines the last public input of the PublicInputsCircuit to be `state_root_prev`, but there's a different value here.
The reason we need `state_root` and `state_root_prev` as public inputs is that these values will be public inputs to the state circuit and will use them to verify the correct transition of the state trie.

The `prev_block_hash` doesn't need to be a public input.

Reply:
Done!